File: projects/01-smart-agriculture/sensors/sensor.py
import json
import os
import random
import time
import urllib.error
import urllib.request
from datetime import datetime, timezone

PROFILES = {
    "soil_moisture":  {"unit": "%",   "lo": 8,  "hi": 45,     "start": 30,    "step": 1.5},
    "temperature":    {"unit": "C",   "lo": 2,  "hi": 42,     "start": 22,    "step": 0.8},
    "humidity":       {"unit": "%",   "lo": 25, "hi": 98,     "start": 60,    "step": 2.0},
    "light_intensity":{"unit": "lux", "lo": 0,  "hi": 100000, "start": 40000, "step": 6000},
    "rainfall":       {"unit": "mm",  "lo": 0,  "hi": 18,     "start": 0,     "step": 3.0},
}


def clamp(value, lo, hi):
    return max(lo, min(hi, value))


def next_value(current, profile):
    drift = random.uniform(-profile["step"], profile["step"])
    return round(clamp(current + drift, profile["lo"], profile["hi"]), 2)


def post_batch(url, payload):
    body = json.dumps(payload).encode()
    req = urllib.request.Request(url, data=body, headers={"Content-Type": "application/json"})
    with urllib.request.urlopen(req, timeout=5) as resp:
        return resp.status


def main():
    sensor_type = os.environ["SENSOR_TYPE"]
    site_id = os.getenv("SITE_ID", "field-1")
    sample_interval = float(os.getenv("SAMPLE_INTERVAL", "2"))
    dispatch_interval = float(os.getenv("DISPATCH_INTERVAL", "10"))
    fog_url = os.getenv("FOG_URL", "http://fog:8000/ingest")

    profile = PROFILES[sensor_type]
    value = profile["start"]
    buffer = []
    last_dispatch = time.monotonic()

    print(f"{sensor_type}@{site_id} sampling every {sample_interval}s, dispatching every {dispatch_interval}s", flush=True)

    while True:
        value = next_value(value, profile)
        buffer.append({"ts": datetime.now(timezone.utc).isoformat(), "value": value})

        if time.monotonic() - last_dispatch >= dispatch_interval and buffer:
            payload = {
                "sensor_type": sensor_type,
                "site_id": site_id,
                "unit": profile["unit"],
                "readings": buffer,
            }
            try:
                post_batch(fog_url, payload)
                print(f"{sensor_type} dispatched {len(buffer)} readings", flush=True)
                buffer = []
                last_dispatch = time.monotonic()
            except urllib.error.URLError as exc:
                print(f"{sensor_type} dispatch failed, will retry: {exc}", flush=True)

        time.sleep(sample_interval)


if __name__ == "__main__":
    main()

File: projects/02-industrial-equipment/sensors/src/main/java/com/fec/industrial/sensor/Sensor.java
package com.fec.industrial.sensor;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;

public class Sensor {

    record Profile(String unit, double lo, double hi, double start, double step) {}

    static final Map<String, Profile> PROFILES = Map.of(
        "vibration",         new Profile("mm/s", 0.2, 9.0, 2.0, 0.4),
        "motor_temperature", new Profile("C", 30, 110, 65, 3.0),
        "bearing_acoustic",  new Profile("dB", 40, 100, 60, 4.0),
        "rotation_speed",    new Profile("RPM", 800, 3600, 1800, 80),
        "power_draw",        new Profile("kW", 5, 75, 35, 5.0)
    );

    static double clamp(double value, double lo, double hi) {
        return Math.max(lo, Math.min(hi, value));
    }

    // Values walk randomly from the previous reading (a bounded random walk)
    // rather than being drawn fresh each tick, so consecutive samples stay
    // close together like a real sensor trace instead of jumping erratically
    // -- this is what makes the fog's windowed min/max/avg and the dashboard
    // sparkline look like plausible equipment telemetry.
    static double nextValue(double current, Profile profile) {
        double delta = ThreadLocalRandom.current().nextDouble(-profile.step(), profile.step());
        double moved = clamp(current + delta, profile.lo(), profile.hi());
        return Math.round(moved * 100.0) / 100.0;
    }

    static String toJson(String sensorType, String siteId, String unit, List<double[]> readings, List<Instant> timestamps) {
        StringBuilder sb = new StringBuilder();
        sb.append("{\"sensor_type\":\"").append(sensorType).append("\",");
        sb.append("\"site_id\":\"").append(siteId).append("\",");
        sb.append("\"unit\":\"").append(unit).append("\",");
        sb.append("\"readings\":[");
        for (int i = 0; i < readings.size(); i++) {
            if (i > 0) sb.append(",");
            sb.append("{\"ts\":\"").append(timestamps.get(i)).append("\",");
            sb.append("\"value\":").append(readings.get(i)[0]).append("}");
        }
        sb.append("]}");
        return sb.toString();
    }

    public static void main(String[] args) throws Exception {
        String sensorType = System.getenv("SENSOR_TYPE");
        if (sensorType == null) throw new IllegalStateException("SENSOR_TYPE env var is required");
        String siteId = System.getenv().getOrDefault("SITE_ID", "line-1");
        double sampleInterval = Double.parseDouble(System.getenv().getOrDefault("SAMPLE_INTERVAL", "2"));
        double dispatchInterval = Double.parseDouble(System.getenv().getOrDefault("DISPATCH_INTERVAL", "10"));
        String fogUrl = System.getenv().getOrDefault("FOG_URL", "http://fog:8000/ingest");

        Profile profile = PROFILES.get(sensorType);
        if (profile == null) throw new IllegalStateException("unknown SENSOR_TYPE: " + sensorType);

        HttpClient client = HttpClient.newHttpClient();
        double value = profile.start();
        List<double[]> buffer = new ArrayList<>();
        List<Instant> timestamps = new ArrayList<>();
        long lastDispatch = System.nanoTime();

        System.out.printf("%s@%s sampling every %ss, dispatching every %ss%n", sensorType, siteId, sampleInterval, dispatchInterval);

        while (true) {
            value = nextValue(value, profile);
            buffer.add(new double[]{value});
            timestamps.add(Instant.now());

            double elapsedSeconds = (System.nanoTime() - lastDispatch) / 1_000_000_000.0;
            if (elapsedSeconds >= dispatchInterval && !buffer.isEmpty()) {
                String body = toJson(sensorType, siteId, profile.unit(), buffer, timestamps);
                try {
                    HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(fogUrl))
                        .header("Content-Type", "application/json")
                        .POST(HttpRequest.BodyPublishers.ofString(body))
                        .timeout(Duration.ofSeconds(5))
                        .build();
                    client.send(request, HttpResponse.BodyHandlers.discarding());
                    System.out.printf("%s dispatched %d readings%n", sensorType, buffer.size());
                    buffer.clear();
                    timestamps.clear();
                    lastDispatch = System.nanoTime();
                } catch (Exception exc) {
                    System.out.printf("%s dispatch failed, will retry: %s%n", sensorType, exc.getMessage());
                }
            }

            Thread.sleep((long) (sampleInterval * 1000));
        }
    }
}
